File: programs/s-rgb-paint/src/ix/paint/burn/two.rs
use anchor_lang::prelude::*;
use partial_application::partial;
use crate::BurnPixelsTwo;
use crate::ix::paint::burn::almost_burn;

pub fn ix(ctx: Context<BurnPixelsTwo>) -> Result<()> {
    // partially apply almost burn
    let burn = partial!(almost_burn =>
        &ctx.accounts.token_program,
        &ctx.accounts.payer,
        &ctx.accounts.proof,
        _, _, _
    );
    // burn pixel one
    burn(
        &ctx.accounts.pixel_one_mint,
        &ctx.accounts.pixel_one_mint_ata,
        |p| &p.burned.plan.one,
    )?;
    // burn pixel two
    burn(
        &ctx.accounts.pixel_two_mint,
        &ctx.accounts.pixel_two_mint_ata,
        |p| &p.burned.plan.two,
    )?;
    // proof
    let proof = &mut ctx.accounts.proof;
    assert_eq!(proof.arity, 2);
    proof.burned.burned = true;
    Ok(())
}

File: web/js/anchor/ix/paint/burn/two.ts
import {AnchorProvider, Program, SplToken} from "@project-serum/anchor";
import {PublicKey, SystemProgram, SYSVAR_RENT_PUBKEY} from "@solana/web3.js";
import * as Proof from "../../../pda/paint/proof-pda";
import * as Pixel from "../../../pda/craft/pixel-pda";
import {deriveAtaPda} from "../../../pda/ata-pda";
import {
    SPL_ASSOCIATED_TOKEN_PROGRAM_ID,
    SPL_TOKEN_PROGRAM_ID
} from "../../../util/constants";
import {SRgbStake} from "../../../idl/stake";
import {SRgbCraft} from "../../../idl/craft";
import {SRgbPaint} from "../../../idl/paint";

export async function ix(
    provider: AnchorProvider,
    programs: {
        stake: Program<SRgbStake>;
        craft: Program<SRgbCraft>;
        paint: Program<SRgbPaint>;
        token: Program<SplToken>
    },
    proof: {
        pda: PublicKey,
        proof: Proof.Proof
    }
): Promise<void> {
    const pixelOne = await Pixel.getPixelPda(
        provider,
        programs,
        proof.proof.burned.plan.one.pda
    );
    const pixelOneMintAta = deriveAtaPda(
        provider.wallet.publicKey,
        pixelOne.mint
    );
    const pixelTwo = await Pixel.getPixelPda(
        provider,
        programs,
        proof.proof.burned.plan.two.pda
    );
    const pixelTwoMintAta = deriveAtaPda(
        provider.wallet.publicKey,
        pixelTwo.mint
    );
    await programs
        .paint
        .methods
        .burnPixelsTwo()
        .accounts(
            {
                proof: proof.pda,
                pixelOne: proof.proof.burned.plan.one.pda,
                pixelTwo: proof.proof.burned.plan.two.pda,
                pixelOneMint: pixelOne.mint,
                pixelOneMintAta: pixelOneMintAta,
                pixelTwoMint: pixelTwo.mint,
                pixelTwoMintAta: pixelTwoMintAta,
                mint: proof.proof.nft.mint,
                payer: provider.wallet.publicKey,
                tokenProgram: SPL_TOKEN_PROGRAM_ID,
                associatedTokenProgram: SPL_ASSOCIATED_TOKEN_PROGRAM_ID,
                systemProgram: SystemProgram.programId,
                rent: SYSVAR_RENT_PUBKEY
            }
        ).rpc()
}
